jgrapht-core/src/main/java/org/jgrapht/generate/BarabasiAlbertGraphGenerator.java

```diff
@@ -0,0 +1,203 @@
+{
+    private final Random rng;
+
+    private final int m0;
+    private final int m;
+    private final int n;
+
+    /**
+     * Constructor
+     * 
+     * @param m0 number of initial nodes
+     * @param m number of edges of each new node added during the network growth
+     * @param n final number of nodes
+     * @throws IllegalArgumentException in case of invalid parameters
+     */
+    public BarabasiAlbertGraphGenerator(int m0, int m, int n)
+    {
+        this(m0, m, n, new Random());
+    }
+
+    /**
+     * Constructor
+     * 
+     * @param m0 number of initial nodes
+     * @param m number of edges of each new node added during the network growth
+     * @param n final number of nodes
+     * @param seed seed for the random number generator
+     * @throws IllegalArgumentException in case of invalid parameters
+     */
+    public BarabasiAlbertGraphGenerator(int m0, int m, int n, long seed)
+    {
+        this(m0, m, n, new Random(seed));
+    }
+
+    /**
+     * Constructor
+     * 
+     * @param m0 number of initial nodes
+     * @param m number of edges of each new node added during the network growth
+     * @param n final number of nodes
+     * @param rng the random number generator to use
+     * @throws IllegalArgumentException in case of invalid parameters
+     */
+    public BarabasiAlbertGraphGenerator(int m0, int m, int n, Random rng)
+    {
+        if (m0 < 1) {
+            throw new IllegalArgumentException("invalid initial nodes (" + m0 + " < 1)");
+        }
+        this.m0 = m0;
+        if (m <= 0) {
+            throw new IllegalArgumentException("invalid edges per node (" + m + " <= 0");
+        }
+        if (m > m0) {
+            throw new IllegalArgumentException("invalid edges per node (" + m + " > " + m0 + ")");
+        }
+        this.m = m;
+        if (n < m0) {
+            throw new IllegalArgumentException(
+                "total number of nodes must be at least equal to the initial set");
+        }
+        this.n = n;
+        this.rng = Objects.requireNonNull(rng, "Random number generator cannot be null");
+    }
+
+    /**
+     * Generates an instance.
+     * 
+     * @param target the target graph
+     * @param vertexFactory the vertex factory
+     * @param resultMap not used by this generator, can be null
+     */
+    @Override
+    public void generateGraph(
+        Graph<V, E> target, VertexFactory<V> vertexFactory, Map<String, V> resultMap)
+    {
+        /*
+         * Ensure directed or undirected.
+         */
+        boolean isDirected;
+        if (target instanceof DirectedGraph<?, ?>) {
+            isDirected = true;
+        } else if (target instanceof UndirectedGraph<?, ?>) {
+            isDirected = false;
+        } else {
+            throw new IllegalArgumentException("Graph must be either directed or undirected");
+        }
+
+        /*
+         * Create complete graph with m0 nodes
+         */
+        List<V> nodes = new ArrayList<>(n * m);
```

Comment:
I first wrote this and then I fixed the `CompleteGraphGenerator` which had some issues (did not work with Pseudographs, etc). Latest version now uses the generator.